File: apps/pulse/tahs/tah_entry_layout.py
from dataclasses import dataclass
import struct
from typing import Union

ENTRY_SIZE = 80
ENTRY_HEADER_SIZE = 24
ENTRY_SPEC_SIZE = 56
ENTRY_PAD_SIZE = 7
SPEC_V1_PAYLOAD_SIZE = 52

_ENTRY_STRUCT = struct.Struct("<B7sQII56s")
_SPEC_V1_STRUCT = struct.Struct("<BBH52s")


BytesLike = Union[bytes, bytearray, memoryview, str]


@dataclass(frozen=True)
class TahSpecV1:
    kind: int
    version: int
    flags: int
    payload: BytesLike = b""


@dataclass(frozen=True)
class TahEntry:
    tag: int
    offset: int
    length: int
    meta: int
    spec: Union[TahSpecV1, BytesLike]
    pad: BytesLike = b""


@dataclass(frozen=True)
class DecodedTahEntry:
    tag: int
    pad: bytes
    offset: int
    length: int
    meta: int
    spec: TahSpecV1
    raw_spec: bytes


def pack_spec_v1(spec: TahSpecV1) -> bytes:
    _assert_uint("spec.kind", spec.kind, 0xFF)
    _assert_uint("spec.version", spec.version, 0xFF)
    _assert_uint("spec.flags", spec.flags, 0xFFFF)
    return _SPEC_V1_STRUCT.pack(
        spec.kind,
        spec.version,
        spec.flags,
        _fixed_bytes(spec.payload, SPEC_V1_PAYLOAD_SIZE, "spec.payload"),
    )


def unpack_spec_v1(data: BytesLike) -> TahSpecV1:
    raw = _exact_bytes(data, ENTRY_SPEC_SIZE, "spec")
    kind, version, flags, payload = _SPEC_V1_STRUCT.unpack(raw)
    return TahSpecV1(kind=kind, version=version, flags=flags, payload=payload)


def pack_entry(entry: TahEntry) -> bytes:
    _assert_uint("entry.tag", entry.tag, 0xFF)
    _assert_uint("entry.offset", entry.offset, 0xFFFFFFFFFFFFFFFF)
    _assert_uint("entry.length", entry.length, 0xFFFFFFFF)
    _assert_uint("entry.meta", entry.meta, 0xFFFFFFFF)

    spec = (
        pack_spec_v1(entry.spec)
        if isinstance(entry.spec, TahSpecV1)
        else _exact_bytes(entry.spec, ENTRY_SPEC_SIZE, "entry.spec")
    )

    return _ENTRY_STRUCT.pack(
        entry.tag,
        _fixed_bytes(entry.pad, ENTRY_PAD_SIZE, "entry.pad"),
        entry.offset,
        entry.length,
        entry.meta,
        spec,
    )


def unpack_entry(data: BytesLike) -> DecodedTahEntry:
    raw = _exact_bytes(data, ENTRY_SIZE, "entry")
    tag, pad, offset, length, meta, raw_spec = _ENTRY_STRUCT.unpack(raw)
    return DecodedTahEntry(
        tag=tag,
        pad=pad,
        offset=offset,
        length=length,
        meta=meta,
        spec=unpack_spec_v1(raw_spec),
        raw_spec=raw_spec,
    )


def _assert_uint(name: str, value: int, maximum: int) -> None:
    if not isinstance(value, int) or value < 0 or value > maximum:
        raise ValueError(f"{name} must be an unsigned integer <= {maximum}")


def _fixed_bytes(value: BytesLike, size: int, label: str) -> bytes:
    raw = _coerce_bytes(value)
    if len(raw) > size:
        raise ValueError(f"{label} must be {size} bytes or less")
    return raw.ljust(size, b"\0")


def _exact_bytes(value: BytesLike, size: int, label: str) -> bytes:
    raw = _coerce_bytes(value)
    if len(raw) != size:
        raise ValueError(f"{label} must be exactly {size} bytes")
    return raw


def _coerce_bytes(value: BytesLike) -> bytes:
    if isinstance(value, str):
        return value.encode("utf-8")
    return bytes(value)

File: apps/pulse/lib/core/tah_entry_layout.ts
import { Buffer } from 'node:buffer';

export const TAH_ENTRY_SIZE = 80;
export const TAH_ENTRY_HEADER_SIZE = 24;
export const TAH_ENTRY_SPEC_SIZE = 56;
export const TAH_ENTRY_PAD_SIZE = 7;
export const TAH_SPEC_V1_PAYLOAD_SIZE = 52;

export interface TahSpecV1 {
  kind: number;
  version: number;
  flags: number;
  payload?: Buffer | Uint8Array | string;
}

export interface TahEntry {
  tag: number;
  pad?: Buffer | Uint8Array;
  offset: bigint | number;
  length: number;
  meta: number;
  spec: TahSpecV1 | Buffer | Uint8Array;
}

export interface DecodedTahEntry {
  tag: number;
  pad: Buffer;
  offset: bigint;
  length: number;
  meta: number;
  spec: TahSpecV1;
  rawSpec: Buffer;
}

export function encodeTahSpecV1(spec: TahSpecV1): Buffer {
  assertUInt('spec.kind', spec.kind, 0xff);
  assertUInt('spec.version', spec.version, 0xff);
  assertUInt('spec.flags', spec.flags, 0xffff);

  const out = Buffer.alloc(TAH_ENTRY_SPEC_SIZE);
  out.writeUInt8(spec.kind, 0);
  out.writeUInt8(spec.version, 1);
  out.writeUInt16LE(spec.flags, 2);
  toFixedBuffer(spec.payload ?? Buffer.alloc(0), TAH_SPEC_V1_PAYLOAD_SIZE, 'spec.payload').copy(out, 4);
  return out;
}

export function decodeTahSpecV1(bytes: Uint8Array): TahSpecV1 {
  const spec = toExactBuffer(bytes, TAH_ENTRY_SPEC_SIZE, 'spec');
  return {
    kind: spec.readUInt8(0),
    version: spec.readUInt8(1),
    flags: spec.readUInt16LE(2),
    payload: spec.subarray(4, TAH_ENTRY_SPEC_SIZE)
  };
}

export function encodeTahEntry(entry: TahEntry): Buffer {
  assertUInt('entry.tag', entry.tag, 0xff);
  assertUInt('entry.length', entry.length, 0xffffffff);
  assertUInt('entry.meta', entry.meta, 0xffffffff);

  const offset = BigInt(entry.offset);
  if (offset < 0n || offset > 0xffffffffffffffffn) {
    throw new RangeError('entry.offset must be an unsigned 64-bit integer');
  }

  const out = Buffer.alloc(TAH_ENTRY_SIZE);
  const pad = entry.pad
    ? toExactBuffer(entry.pad, TAH_ENTRY_PAD_SIZE, 'entry.pad')
    : Buffer.alloc(TAH_ENTRY_PAD_SIZE);
  const spec = entry.spec instanceof Uint8Array
    ? toExactBuffer(entry.spec, TAH_ENTRY_SPEC_SIZE, 'entry.spec')
    : encodeTahSpecV1(entry.spec);

  out.writeUInt8(entry.tag, 0);
  pad.copy(out, 1);
  out.writeBigUInt64LE(offset, 8);
  out.writeUInt32LE(entry.length, 16);
  out.writeUInt32LE(entry.meta, 20);
  spec.copy(out, TAH_ENTRY_HEADER_SIZE);
  return out;
}

export function decodeTahEntry(bytes: Uint8Array): DecodedTahEntry {
  const entry = toExactBuffer(bytes, TAH_ENTRY_SIZE, 'entry');
  const rawSpec = entry.subarray(TAH_ENTRY_HEADER_SIZE, TAH_ENTRY_SIZE);

  return {
    tag: entry.readUInt8(0),
    pad: entry.subarray(1, 8),
    offset: entry.readBigUInt64LE(8),
    length: entry.readUInt32LE(16),
    meta: entry.readUInt32LE(20),
    spec: decodeTahSpecV1(rawSpec),
    rawSpec
  };
}

function assertUInt(name: string, value: number, max: number) {
  if (!Number.isInteger(value) || value < 0 || value > max) {
    throw new RangeError(`${name} must be an unsigned integer <= ${max}`);
  }
}

function toFixedBuffer(value: Buffer | Uint8Array | string, size: number, label: string): Buffer {
  const source = typeof value === 'string' ? Buffer.from(value, 'utf8') : Buffer.from(value);
  if (source.length > size) {
    throw new RangeError(`${label} must be ${size} bytes or less`);
  }

  const out = Buffer.alloc(size);
  source.copy(out);
  return out;
}

function toExactBuffer(value: Uint8Array, size: number, label: string): Buffer {
  const source = Buffer.from(value);
  if (source.length !== size) {
    throw new RangeError(`${label} must be exactly ${size} bytes`);
  }
  return source;
}
